include/zephyr/bluetooth/audio/hap.h

```diff
@@ -0,0 +1,394 @@
+struct bt_hap_harc;
+
+/** Hearing Aid device Info Structure */
+struct bt_hap_harc_info {
+	/** Hearing Aid Type. */
+	enum bt_has_hearing_aid_type type;
+	/** Hearing Aid Features. */
+	enum bt_has_features features;
+	/** Hearing Aid Capabilities. */
+	enum bt_has_capabilities caps;
+	struct {
+		struct bt_hap_harc *pair;
+		const struct bt_csip_set_coordinator_csis_inst *csis;
+		const struct bt_csip_set_coordinator_set_member *member;
+	} binaural;
+};
+
+/** @brief Hearing Aid Remote Controller callback structure. */
+struct bt_hap_harc_cb {
+	/**
+	 * @brief Profile has been unbound.
+	 *
+	 * This callback notifies the application that the profile has been unbound.
+	 *
+	 * @param conn Controller instance pointer.
+	 */
+	void (*unbound)(struct bt_hap_harc *harc);
+
+	/**
+	 * @brief Controller has been connected.
+	 *
+	 * This callback notifies the application of a new connection to remote Hearing Aid device.
+	 * In case the err parameter is non-zero it means that the connection establishment failed.
+	 *
+	 * @param conn Controller instance pointer.
+	 * @param err 0 in case of success or negative errno or ATT Error code in case of error.
+	 */
+	void (*connected)(struct bt_hap_harc *harc, int err);
+
+	/**
+	 * @brief Controller has been disconnected.
+	 *
+	 * This callback notifies the application that the controller has been disconnected.
+	 *
+	 * @param conn Controller instance pointer.
+	 */
+	void (*disconnected)(struct bt_hap_harc *harc);
+
+	/**
+	 * @brief Controller info has been updated.
+	 *
+	 * This callback notifies the application that the controller info has been changed,
+	 * e.g. the Hearing Aid Features has been updated.
+	 *
+	 * @param conn Controller instance pointer.
+	 * @param info Info object.
+	 */
+	void (*info)(struct bt_hap_harc *harc, const struct bt_hap_harc_info *info);
+
+	sys_snode_t _node;
+};
+
+/**
+ * @brief Register Hearing Aid Remote Controller callbacks.
+ *
+ * @param cb The callback structure.
+ *
+ * @return 0 if success, errno on failure.
+ */
+int bt_hap_harc_cb_register(struct bt_hap_harc_cb *cb);
+
+/**
+ * @brief Unregister Hearing Aid Remote Controller callbacks.
+ *
+ * @param cb The callback structure.
+ *
+ * @return 0 if success, errno on failure.
+ */
+int bt_hap_harc_cb_unregister(struct bt_hap_harc_cb *cb);
+
+/**
+ * @brief Bind profile.
+ *
+ * Bind profile to remote device implementing HAP Hearing Aid role.
+ *
+ * @param conn Bluetooth connection object.
+ * @param[out] harc Valid controller object on success.
+ *
+ * @return Zero on success or (negative) error code on failure.
+ */
+int bt_hap_harc_bind(struct bt_conn *conn, struct bt_hap_harc **harc);
+
+/**
+ * @brief Unbind profile.
+ *
+ * Unbind profile or cancel pending profile binding.
+ *
+ * @param harc Controller instance pointer.
+ *
+ * @return Zero on success or (negative) error code on failure.
+ */
+int bt_hap_harc_unbind(struct bt_hap_harc *harc);
```

Comment:
The term or concept of `bind` does not really exist anywhere else in Zephyr, and it is not explained here what it means. 

Can you elaborate? Is this an alternative to the `discover` functions we have for other profiles?

Reply:
Yes, I broke the compatibility with other profiles by adding the `bind`/`unbind` functions.
The reason is there's a need to give application possibility to clear (unbind) the instance. That includes GATT subscription removal and removal of service-related persistent storage.

Reply:
> That includes GATT subscription removal 

Why is that a useful feature?

> removal of service-related persistent storage.

You could argue that removing the bond for the device should remove this information, similar to how we handle keys. With this approach, when removing a bonded device, we would have to call the `_unbind` for each profile that implements it. The profile implementation could use the `bt_conn_auth_info_cb.bond_deleted` callback to trigger this behavior; unless you have some other usecase in mind where you'd only want to clear e.g. HAP data?

Reply:
IMO, Ideally, the application should not be bothered with these details.  However, I can also see how ownership of persistent storage might lie with the application, which might require an application level cleanup.  Is this something we can fix in a subsequent PR?

Reply:
> However, I can also see how ownership of persistent storage might lie with the application, which might require an application level cleanup. Is this something we can fix in a subsequent PR?

That should be done via a callback if it is needed IMO